### ISSUE TITLE
Azure functions will reinitialize Spring Context on each request

### ISSUE THREAD

**Issue by aropop** [spring-cloud/spring-cloud-function#425]
Hi,

I've upgraded from Greenwhich to Hoxton and now each request will trigger the spring context to be initialized. 

I've investigated this issue and found that in Greenwhich, the AzureSpringFunctionInitializer would store the application context in a static property. However in Hoxton this functionality seems to be removed.

```
public class AzureSpringFunctionInitializer implements Closeable {
	private volatile static ConfigurableApplicationContext context;

...

	protected void initialize(ExecutionContext ctxt) {

		ConfigurableApplicationContext context = AzureSpringFunctionInitializer.context;

...
		if (context == null) {
			synchronized (AzureSpringFunctionInitializer.class) {
				if (context == null) {
					ClassUtils.overrideThreadContextClassLoader(
							AzureSpringFunctionInitializer.class.getClassLoader());
					context = springApplication().run();
					AzureSpringFunctionInitializer.context = context;
				}
			}

		}
...
```

For reference: 
* I'm not using the functional bean definition
*  I've also updated to Spring Boot 2.2
* Testing is done locally with latest azure java worker runtime

**Comment by Azer1652** [comment]
I'm having the same issue since Hoxton.

**Comment by aropop** [comment]
@olegz Your solution is indeed more elegant than #428, however it will inject the same execution context into each request. Which might cause bugs, since it might provide for example the wrong function name when context in injected?

Since it is a breaking change it might be interesting to consider a solution similar to #428

**Comment by olegz** [comment]
@aropop Unfortunately I can not willingly introduce any breaking changes in the point release.
That said, I have just pushed some additional improvements which I believe address your latest comment - https://github.com/spring-cloud/spring-cloud-function/commit/c36d1b8594ee52c457db0beb375f29d4adbb3cd4

In any event, thank you for quick feedbacks

**Comment by aropop** [comment]
@olegz I think you might have misunderstood my point. In the method registerTargetContext you register the current TargetContext when the function is initialized. However, on the next invocation, this context is different (at least in the case of azure). But since the first context is already in the spring context, the actual correct TargetContext will not be injected correctly on consequetive requests. The TargetContext (ExecutionContext for the azure case) also contains invocationId, which is unique for each request.

I think your last commit c36d1b8 will reintroduce the original issue, as the spring context will reinitialize on each different function call.
For example if I have an azure function application with 3 different functions A, B, C and 1 spring configuration. If A is triggered and then B is triggered the spring context will have reinitialised, while the main class is the same.

I understand that you cannot introduce any breaking change into this release. But I think that the last commit makes matters worse than they were.

**Comment by olegz** [comment]
@aropop unless I am completely missing something. . .
You stated that upon the next invocation `ExecutionContext` may request a different function. So the last commit caches the previous function name in the static variable similar to the `thisInitializer`. And then the following IF statement gets invoked
```
if (thisInitializer == null || !thisInitializer.functioinName.equals(name)) {
     // reinitialize
else {
    // go the same route as commit the one before last- https://github.com/spring-cloud/spring-cloud-function/commit/7e0e8ff7ee7e7c831e3111f0b1e4b0cf9409beb8
}
```

So all I did is kept the previous fix and then added a condition upon which it will reinitialise if function name is different.

Yes, I do agree that there may be something else that is different (like the invocation id) in the `ExecutionContext`, and I'll see what I can do with that, but I am not sure this makes maters worse. 

Basically it goes as before and will only reinitialise on different function name.

Please let me know ASAP your thoughts as we are about to kick off 3.0.1.

**Comment by olegz** [comment]
Also, from the standpoint of spring-cloud-function we are not and haven't been doing anything with invocation id. So perhaps an enhancement/feature request (e.g., incorporate it into a Message header), but that would be post 3.0.1.

**Comment by aropop** [comment]
@olegz This is indeed how I understood it aswell.

But in azure (I'm not familiar with AWS), you can have an application with multiple functions. Each function has a different name. But each function is defined in the same spring context. 

Take application A with following structure
`
-- SampleFunctionApplication 
--- FunctionWithNameA
--- FunctionWithNameB
`

If function A is triggered the spring context is initialized and the static field is set to A. If then function B is triggered the function names will not match and the spring context will be reinitialized, while it should not, since the SampleFunctionApplication is the only existing Spring configuration/application

**Comment by olegz** [comment]
Ok, that is indeed correct. But for now that is the price I am willing to pay since generally we assume and kind of designed for a single function. 

That said, I will raise a separate enhancement issue and will link to this one so we can address it for 3.1.

Thank you for all your help on this one!

**Comment by olegz** [comment]
Here is the new issue - https://github.com/spring-cloud/spring-cloud-function/issues/437
